### PR TITLE
wgengine/router: add a firewall-based "default route killswitch".

### PR DESCRIPTION
**Pull request by danderson**
Empirically, this is the only reliable way to force Windows to switch
over cleanly to a higher-priority default route: block the old one
such that the sockets get I/O errors and are forced to redial.

The code is copied and adapted from wireguard-windows's firewall module,
with modifications to adjust to Tailscale: we condition only on process
name (same as our existing firewall settings), and we allow DNS out
through the old default route (because right now it's tricky for us
to figure out the correct DNS servers to permit).

This is a bit of a temporary hack. The technique is sound, but we should
be doing more to enable smooth switchover in both directions. However,
this change gets us from default routes not working correctly, to
default routes working mostly correctly. Followups will make it cleaner.

Signed-off-by: David Anderson <danderson@tailscale.com>

### CODE REVIEW

**Comment by danderson** [comment]
I apologize for this code. It's the minimum necessary change to make default routes shippable, but it's not pretty. I'm working on making it prettier as a followup.

**Comment by danderson** [comment]
Go mod is mad at the windows-only package. Tried adding windows build constraints to all the files, but the error remained the same. Wisdom pls? What am I doing wrong?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:120]: hasDefault

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:121]: I'd pull this into a method personally, but your call. you can keep the variable too since you use it twice below.

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:129]: Reading this code for the first time and I have no clue what `false, nil` means here. Pull out into variables? Or add `/* comments */` or `// comments`?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:33]: this is the only mutable field, right? separate out by a blank link and document the concurrency rules? owned by Set? which is not called concurrently due to .... ?

File: scripts/check_license_headers.sh
Comment (bradfitz) [inline scripts/check_license_headers.sh:41]: tabs/spaces mismatch with rest of file

File: wgengine/router/firewall/README.md
Comment (bradfitz) [inline wgengine/router/firewall/README.md:1]: I can see wanting to patch the name package name "firewall" from upstream, but putting it under "wgengine/router" makes it sound like a portable thing available on different OSes.

Maybe `wgengine/router/windows/firewall` or `tempfork/windows-windows/firewall`?

**Review by bradfitz — COMMENTED** [review]
> Go mod is mad at the windows-only package. Tried adding windows build constraints to all the files, but the error remained the same. Wisdom pls? What am I doing wrong?

The problem is that a a bunch of checks do ` ./...` but that expands to include the new firewall package which is not restricted to only Windows. For example:

```
$ go list ./... | grep firewall
tailscale.com/wgengine/router/firewall
$ go install ./wgengine/router/firewall
package tailscale.com/wgengine/router/firewall
        imports golang.org/x/sys/windows: build constraints exclude all Go files in /home/bradfitz/go/pkg/mod/golang.org/x/sys@v0.0.0-20210220050731-9a76102bfb43/windows
```

**Comment by bradfitz** [comment]
In particular, these are your problem files that need build tags:

```
$ go list -f {{.GoFiles}} ./wgengine/router/firewall 
[blocker.go helpers.go mksyscall.go rules.go types_windows_64.go]
```

**Comment by bradfitz** [comment]
```patch
$ git di
diff --git a/wgengine/router/firewall/blocker.go b/wgengine/router/firewall/blocker.go
index 5acfc45..e69d791 100644
--- a/wgengine/router/firewall/blocker.go
+++ b/wgengine/router/firewall/blocker.go
@@ -1,3 +1,5 @@
+// +build windows
+
 /* SPDX-License-Identifier: MIT
  *
  * Copyright (C) 2019-2021 WireGuard LLC. All Rights Reserved.
diff --git a/wgengine/router/firewall/helpers.go b/wgengine/router/firewall/helpers.go
index e3e4eac..1b578e1 100644
--- a/wgengine/router/firewall/helpers.go
+++ b/wgengine/router/firewall/helpers.go
@@ -1,3 +1,5 @@
+// +build windows
+
 /* SPDX-License-Identifier: MIT
  *
  * Copyright (C) 2019-2021 WireGuard LLC. All Rights Reserved.
diff --git a/wgengine/router/firewall/rules.go b/wgengine/router/firewall/rules.go
index e12d388..72fa20a 100644
--- a/wgengine/router/firewall/rules.go
+++ b/wgengine/router/firewall/rules.go
@@ -1,3 +1,5 @@
+// +build windows
+
 /* SPDX-License-Identifier: MIT
  *
  * Copyright (C) 2019-2021 WireGuard LLC. All Rights Reserved.
diff --git a/wgengine/router/firewall/types_windows_64.go b/wgengine/router/firewall/types_windows_64.go
index 6e60aa5..2f20fa8 100644
--- a/wgengine/router/firewall/types_windows_64.go
+++ b/wgengine/router/firewall/types_windows_64.go
@@ -1,4 +1,5 @@
 // +build amd64 arm64
+// +build windows
 
 /* SPDX-License-Identifier: MIT
  *
```

and it's good

File: wgengine/router/router_windows.go
Comment (danderson) [reply wgengine/router/router_windows.go:129]: Pulled into variables. One of them is unused and looks silly, but it's with the aim of minimizing the diff from the imported firewall package.

File: wgengine/router/router_windows.go
Comment (danderson) [reply wgengine/router/router_windows.go:33]: Separated and documented that wgengine serializes calls to router.

File: wgengine/router/firewall/README.md
Comment (danderson) [reply wgengine/router/firewall/README.md:1]: Moved under tempfork. I forgot we have that very useful subdir.

**Comment by danderson** [comment]
Build issues should be fixed, waiting on CI.

As discussed in eng meeting, I'm going to move this into running as a subprocess to defend against potential unsafety problems. Prepping that change now, which will be an extra commit on top of this change. PTAL to expedite followup review?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:142]: bool {

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:128]: I imagine you did `go install ./wgengine/router` on Linux :)

Comment (bradfitz) [reply wgengine/router/router_windows.go:128]: var unusedT *firewall.T

Comment (danderson) [reply wgengine/router/router_windows.go]: Sigh, yes. "it compiles on linux" is not a good check when doing windows stuff, turns out. Fixed so tailscaled actually builds.

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:129]: unused what, though? How about `var allowedIPs []net.IP // none` or something? (I'm making up "allowed IPs")

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:36]: this says who owns it but not what it is.

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:130]: maybe drop some duration measurements around these calls like we do for netsh so we can see how long these are taking in the field?

**Comment by bradfitz** [comment]
Can you break this commit into two commits: first adding the tempfork package, and then using it?

**Comment by danderson** [comment]
Tempfork moved to its own commit, main "use" commit spawns and manages a subprocess. Tested on my windows VM, killswitch still works.

**Comment by bradfitz** [comment]
> Tempfork moved to its own commit

Please include archaeological info in the commit message: which upstream git hash did the code come from?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:317]: ft.logf?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:235]: "a, b, c bool"

also, worth documenting this signature a bit? lot of stuff going on.

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:309]: NewReaderSize(out, 1<<10) or less? I guess it's Windows so it's fine, but I'm used to counting pennies on iOS.

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:164]: document the lifetime of this Cmd? this can come & go & come back based on config, right?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:177]: signature changed but docs didn't. how does routes come into play?

File: wgengine/router/router_windows.go
Comment (danderson) [reply wgengine/router/router_windows.go:235]: The bools aren't great, but doesn't seem worth making an options struct given that I'm planning to replace most of this code with WFP API things after 1.6.

Added more documentation. The main purpose of doSet is to be a func that can return an error for BackOff in doAsyncSet, rather than have to do subtle breaks and gotos to get the control flow right.

File: wgengine/router/router_windows.go
Comment (danderson) [reply wgengine/router/router_windows.go:309]: Sure, why not.

**Comment by danderson** [comment]
Archaeology information added.

File: wgengine/router/router_windows.go
Comment (bradfitz) [reply wgengine/router/router_windows.go:249]: I'm not saying write a struct. Just to merge multiple args of the same type:

```patch
- func (ft *firewallTweaker) doSet(local []string, killswitch bool, clear bool, procRule bool) error {
+ func (ft *firewallTweaker) doSet(local []string, killswitch, clear, procRule bool) error {
```

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:331]: using just "%s" makes this rate-limited along with all other "%s" in the program (which Josh has been on a mission to eliminate)

How about "fw-child: %s"?

File: wgengine/router/router_windows.go
Comment (bradfitz) [inline wgengine/router/router_windows.go:319]: out.Close() before the return

**Review by bradfitz — APPROVED** [review]

**Pull request merged**
